3D: Auto-adjust date ranges and remove 1-year limit; add data range API

hmhngx opened the pull request:
# Description

(Please include a summary of the change and which issue is touched on. Please also include relevant motivation and context.)

This PR enhances the 3D visualization experience and supporting APIs:
- Adds backend data range API endpoints to support 3D auto-adjustment.
- Implements client-side utilities to auto-adjust 3D graphics based on available data.
- Enables 3D date ranges greater than one year by removing the previous one-year limit.
- Minor cleanup and comment updates in related UI components.

Fixes #1185 

(In general, OED likes to have at least one issue associated with each pull request. Replace [issue] with the OED GitHub issue number. In the preview you will see an issue description if you hover over that number. You can create one yourself before doing this pull request. This is where details are normally given on what is being addressed. Note you should not use the word "Fixes" if it does not completely address the issue since the issue would automatically be closed on merging the pull request. In that case use "Partly Addresses #[issue].)

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations
- API and UI changes should be validated across large datasets for performance.
- Additional integration tests may be needed for edge cases around very sparse or extremely long time ranges.
- Documentation for the new data range endpoints and 3D auto-adjustment behavior should be added in a follow-up if not covered by existing docs.

(Describe any issues that remain or work that should still be done.)

hmhngx committed:
feat: add dataRange endpoint for 3D auto-adjustment

hmhngx committed:
feat: add auto-adjustment for 3D date ranges > 1 year

hmhngx committed:
feat: add auto-adjustment utilities for 3D graphics

hmhngx committed:
feat: add data range API endpoints and remove 3D 1-year limit

hmhngx committed:
feat: add numDays and threeDInterval to Redux state

- Add numDays to ThreeDState for user-selectable date range span
- Add threeDInterval to GraphState for calculated 3D date range
- Add reducers and selectors for both fields
- Support numDays in chart link parsing

hmhngx committed:
feat: implement 4-case 3D date range calculation logic

- Add MAX_3D_DAYS (1095) and DEFAULT_3D_DAYS (365) constants
- Implement calculateThreeDDateRange() with 4-case logic:
  * Bounded/Bounded: warn if > maxDays, else use picker values
  * Unbounded/Bounded: use [DRE - numDays, DRE]
  * Bounded/Unbounded: use [max(DRS, end-numDays), latest day]
  * Unbounded/Unbounded: use [end - numDays, latest day]
- Update isValidThreeDInterval() to accept maxDays parameter

hmhngx committed:
feat: add ThreeDIntervalControlsComponent for numDays selection

- Dropdown with options: 1 month, 6 months, 1 year, 2 years, max, custom
- Custom input with validation (1-1095 days)
- Updates Redux state on selection
- Similar pattern to IntervalControlsComponent

hmhngx committed:
feat: integrate numDays into ThreeDComponent

- Calculate 3D date range using 4-case logic in useEffect
- Use dataRange API to get available data dates
- Store calculated interval in Redux (threeDInterval)
- Show warning when date range exceeds maxDays
- Use threeDInterval for 3D queries instead of queryTimeInterval
- Add ThreeDIntervalControlsComponent to UI options
- Fix infinite loop by only dispatching on interval change

hmhngx committed:
feat: add chart link support and translations for numDays

- Include numDays in 3D chart link generation
- Add translation keys: threeD.numDays, threeD.date.range.exceeds.max
- Add translations for 6.months and 2.years
- Support English, French, and Spanish

huss commented:
~The type check failed in the build and should be addressed.~

huss commented on src/client/app/redux/api/readingsApi.ts:
```diff
@@ -150,6 +150,10 @@ export const readingsApi = baseApi.injectEndpoints({
 			},
 			providesTags: ['Readings']
+		}),
+		dataRange: builder.query<{ minDate: string | null, maxDate: string | null }, { id: number, meterOrGroup: 'meters' | 'groups' }>({
```
I'm not sure but can the type of meterOrGroup be set by the enum MeterOrGroup in src/client/app/types/redux/graph.ts?

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -494,25 +486,80 @@ function createRouter() {
+			const meterIDs = req.params.meter_ids.split(',').map(idStr => Number(idStr));
+			const conn = getConnection();
+			try {
+				const result = await conn.oneOrNone(`
```
The SQL query should probably be in the Meter model and the SQL in its own file as with most queries. Same for groups.

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -494,25 +486,80 @@ function createRouter() {
+				const groupID = req.params.group_id;
+				const graphicUnitID = req.query.graphicUnitId;
+				const readingInterval = req.query.readingInterval;
+				const forJson = await groupThreeDReadings(groupID, graphicUnitID, timeInterval, readingInterval);
```
This is an open route (login not needed). Thus, a bad actor could request arbitrary readingInterval and return a lot of data. Before it limited it to 1 year but that is no longer done. The range needs to be limited to the desired value. Applies to meter too.

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -494,25 +486,80 @@ function createRouter() {
+						MIN(start_timestamp) as min_date,
+						MAX(end_timestamp) as max_date
+					FROM readings 
+					WHERE meter_id = ANY($1)
```
I was thinking that the limit would be for each meter and now across all meters (or groups) since you can only graph one at a time. It makes me wonder if this route and related code should only be for a single data source. I'm not sure how that would impact the code.

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -494,25 +486,80 @@ function createRouter() {
+					WHERE meter_id = ANY($1)
+				`, [meterIDs]);
+				
+				if (result && result.min_date && result.max_date) {
```
Does this only happen if there are no readings for the data source? Just wondering.

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -494,25 +486,80 @@ function createRouter() {
+					res.json({ minDate: null, maxDate: null });
+				}
+			} catch (error) {
+				console.error('Error getting meter data range:', error);
```
All errors must be logged and not to the console. Better yet, use the success/failure methods found in src/server/services/csvPipeline/success.js.

huss commented on src/client/app/components/ThreeDComponent.tsx:
```diff
@@ -46,6 +49,68 @@ export default function ThreeDComponent() {
+		} else if (!result.shouldWarn) {
+			prevWarningRef.current = false;
+		}
+	}, [meterOrGroupID, dataRange, queryTimeInterval, numDays, dispatch, threeDMeterOrGroup, translate]);
```
I'm unclear on how dispatch or translate change.

huss commented on src/client/app/components/ThreeDComponent.tsx:
```diff
@@ -46,6 +49,68 @@ export default function ThreeDComponent() {
+	const prevCalculatedIntervalRef = React.useRef<string | null>(null);
+
+	// Calculate 3D date range based on the 4-case logic
+	React.useEffect(() => {
```
In the big picture, I'm wondering about the use of a Redux selector vs having the logic here.

At an even higher level, I'm thinking if all the logic should be on the server. My thinking is the user selects the date range and the span and then the result is based on data on the server. This is why I'm thinking it may be easier and better to do the calculation on the server. This might address other comments about bad routes, the useEffect, etc. I'm open to thoughts.

huss commented on src/client/app/components/ThreeDComponent.tsx:
```diff
@@ -8,22 +8,24 @@ import { selectGroupDataById } from '../redux/api/groupsApi';
 import { selectThreeDComponentInfo } from '../redux/selectors/threeDSelectors';
 import { selectScalingFromEntity } from '../redux/selectors/entitySelectors';
 import { selectGraphState } from '../redux/slices/graphSlice';
+import { updateTimeInterval, updateThreeDInterval, selectThreeDNumDays, selectQueryTimeInterval, selectThreeDMeterOrGroup } from '../redux/slices/graphSlice';
```
There are a couple of lines longer than 150 characters. ESLint can show you where they are.

huss commented on src/client/app/components/ThreeDComponent.tsx:
```diff
@@ -46,6 +49,68 @@ export default function ThreeDComponent() {
+			prevCalculatedIntervalRef.current = intervalString;
+		}
+
+		// Show warning to user if result.shouldWarn is true (only once per change)
```
I have not looked into the details but having to check for pervious messages makes me think there may be a bigger picture issue in the logic. This may go away so I'm not thinking about it too much.

huss commented on src/client/app/utils/dateRangeCompatibility.ts:
```diff
@@ -100,10 +100,133 @@ export function roundTimeIntervalForFetch(timeInterval: TimeInterval): TimeInter
 }
 
 /**
```
I'm not going to review these changes until the issue about if it is client or server code is resolved.

huss replied:
I'm still unsure that the calculation should happen on the client. I think we should discuss to resolve. Then I can carefully review the code.

huss commented on src/client/app/components/ThreeDIntervalControlsComponent.tsx:
```diff
@@ -0,0 +1,187 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This file seems to have a high overlap with src/client/app/components/IntervalControlsComponent.tsx. Given that file already has logic for different chart types, I'm hoping it would be better to use the other component where it is modified to deal with the new 3D cases.

huss commented on src/client/app/components/ThreeDIntervalControlsComponent.tsx:
```diff
@@ -0,0 +1,187 @@
+		SixMonths: '6.months',
+		OneYear: '1.year',
+		TwoYears: '2.years',
+		Max: 'max',
```
I think from a UI perspective and to be consistent with other menus, this should show '3 years'. I realize it means that changes might mean editing but I'm also hoping that doing the translation as in src/client/app/components/IntervalControlsComponent.tsx will help mitigate that. Also, bar does the translation in a somewhat different way.

huss commented on src/client/app/components/ThreeDIntervalControlsComponent.tsx:
```diff
@@ -0,0 +1,187 @@
+import { useTranslate } from '../redux/componentHooks';
+import TooltipMarkerComponent from './TooltipMarkerComponent';
+import { labelStyle, divTopBottomPadding } from '../styles/modalStyle';
+import { MAX_3D_DAYS, DEFAULT_3D_DAYS, getEffectiveNumDays } from '../utils/dateRangeCompatibility';
```
DEFAULT_3D_DAYS ia not used. This and a couple of other TS issues are seen in the checks that are failing on the PR.

huss commented on src/client/app/redux/selectors/chartQuerySelectors.ts:
```diff
@@ -194,17 +194,21 @@ export const selectMapChartQueryArgs = createSelector(
 			meterOrGroup: threeD.meterOrGroup!
 		};
-		const shouldSkipQuery = !threeD.meterOrGroupID || !queryTimeInterval.getIsBounded();
+		// Skip query if no meter/group selected or if interval is not bounded
```
I think the issue specified that the 3D graph happens if unbounded as from the last reading time backward. This seems to stop 3D graphics in this case.

huss commented on src/client/app/redux/slices/graphSlice.ts:
```diff
@@ -37,7 +38,8 @@ const defaultState: GraphState = {
 		meterOrGroup: undefined,
-		readingInterval: ReadingInterval.Hourly
+		readingInterval: ReadingInterval.Hourly,
+		numDays: undefined // Default to undefined, will use 1 year (365 days) if not set
```
OED really prefers comments on the line above the code.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -500,12 +500,16 @@ const LocaleTranslationData = {
 		"threeD.time": "Time",
 		"threeD.x.axis.label": "Hours of Day",
 		"threeD.y.axis.label": "Days of Calendar Year",
+		"6.months": "6 months",
```
This is in the wrong place alphabetically. See 2.months for location.

huss commented on src/client/app/types/redux/graph.ts:
```diff
@@ -54,6 +54,7 @@ export interface ThreeDState {
 	meterOrGroupID: MeterOrGroupID | undefined;
 	meterOrGroup: MeterOrGroup | undefined;
 	readingInterval: ReadingInterval;
+	numDays: number | undefined; // Number of days for 3D graphic span (undefined means use default: 1 year)
```
I'm curious about using undefined vs setting the default value to 1 year in src/client/app/redux/slices/graphSlice.ts.

huss reviewed (changes requested):
Thanks to @hmhngx for another contribution. I have made some comments to consider. I also have some thoughts that were not easy to put in a specific place so I'm putting them here.

- The custom range (and maybe the non-custom ones but not checked) seems to include one too many days. For example, choosing 3 days seems to show 4 on the graphic.
  - Update: I'm having problems testing this as the date range is often not honored. This applies to several other ones in the comment.
- Setting the date range outside the range of readings caused a server error and it never graphed (``ERROR:  range lower bound must be less than or equal to range upper bound``) and it then went to the client.
  - Update: This now says rendering and also pops up a warning about the range.
- Some actions are causing a client message of exceeding the max days when it seems valid but this might naturally go away with other items.
- If set the upper bound on the date range that is above the actual readings available then get client message about exceeding days but still graphs the result.
- ~Maybe change ``		"threeD.y.axis.label": "Days of Calendar Year",`` on the 3D graphic label for the axis to "Date" since no longer just one year.~
- I'm thinking about whether the behavior should be changed from the issue if have only DRS is set. In this case, it would then do DRS to DRS + numDays. If there is not enough data in readings then the graphic would end early. This allows a user to easily get data from a given date forward. This also seems parallel to only have the DRE. What do you think?

As always, I'm happy to clarify and discuss as desired.

hmhngx committed:
Add data range endpoints for 3D auto-adjustment

- Add Reading.getMeterDataRange() and Reading.getGroupDataRange() methods
- Add /dataRange/meters/:meter_id and /dataRange/groups/:group_id routes
- Fix validation functions to accept single meter/group IDs
- Fix day calculation mismatch: use diff('days') + 1 to match client calculation
- Fix validateGroupThreeDReadingsParams to check group_id instead of meter_ids
- Add proper error logging for data range requests

hmhngx committed:
Add dataRange API endpoint for 3D graphics

- Add dataRange query endpoint to readingsApi
- Fix meterOrGroup type to use MeterOrGroup enum instead of string literal
- Endpoint fetches min/max timestamps for meter or group data

hmhngx committed:
Fix 3D date range calculation bugs

- Fix 'one too many days' bug: subtract (numDays - 1) for correct calendar days
- Add validation to prevent server errors when range exceeds available data
- Ensure threeDStartDate >= minDataDate and threeDEndDate <= maxDataDate
- Fix calculateThreeDDateRange to handle edge cases with data boundaries
- Improve autoAdjustThreeDInterval validation for maxDataDate

hmhngx committed:
Fix 3D component TypeScript errors and query logic

- Fix undefined args error in selectThreeDQueryArgs selector
- Return undefined args when shouldSkipQuery is true
- Add non-null assertion for args in useQuery (safe when skip=true)
- Remove unused updateTimeInterval import
- Fix comment placement for numDays in graphSlice (OED style preference)

hmhngx committed:
Fix date picker error display for unbounded intervals

- Provide default 30-day range when queryTimeInterval is unbounded
- Prevents DateRangePicker from showing error state (red highlighting)
- Fix TimeInterval import path to use correct relative path
- Add non-null assertion for args in ReadingsPerDaySelectComponent

hmhngx committed:
Update 3D graphics translations and UI labels

- Change 'max' to '3 years' in 3D date range span dropdown
- Add '3.years' translation key in English, French, and Spanish
- Change y-axis label from 'Days of Calendar Year' to 'Date'
- Fix alphabetical order of translation keys (2.years, 3.years, 6.months)
- Remove unused getLabelForValue function

huss commented on src/client/app/components/DateRangeComponent.tsx:
```diff
@@ -35,6 +37,14 @@ export default function DateRangeComponent() {
+	const displayInterval = queryTimeInterval.getIsBounded()
+		? queryTimeInterval
+		: new TimeInterval(
+			moment().subtract(30, 'days').startOf('day'),
```
This seems to be causing several undesirable behaviors. 

1. The date range is shown and set to the thirty days back from the current day. However, the actual 3D data goes back 1 year from the last day with data. Thus, they do not correspond to each other.
2. In src/client/app/components/ThreeDIntervalControlsComponent.tsx there is DEFAULT_3D_DAYS. How does this relate to the 30 days here? This relates to item 1.
3. It is directly setting the data range shown and not setting queryTimeInterval. As a result, the date range is set but not shown on several other graphic pages. It also is not honored.
4. I'm seeing warning, it not honoring the time interval & resetting the time interval in 3D.

My current thinking is the date range should not be modified and the 3D data is gotten from the server by going back by DEFAULT_3D_DAYS (currently 1 year). Thus, date range remains unbounded but 3D gets the correct data. I think it may be possible that this was causing an error given your comment. If so, maybe we should discuss options.

huss commented on src/client/app/redux/selectors/chartQuerySelectors.ts:
```diff
@@ -194,17 +194,25 @@ export const selectMapChartQueryArgs = createSelector(
 	selectSelectedUnit,
 	selectThreeDState,
-	(queryTimeInterval, selectedUnit, threeD) => {
+	(queryTimeInterval, threeDInterval, selectedUnit, threeD) => {
```
queryTimeInterval seems unused.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -12,6 +12,9 @@ const LocaleTranslationData = {
 		"1.week": "1 week",
 		"1.year": "1 year",
 		"2.months": "2 months",
+		"2.years": "2 years",
```
Moving the items does not appear in other languages.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -500,12 +503,14 @@ const LocaleTranslationData = {
 		"threeD.time": "Time",
 		"threeD.x.axis.label": "Hours of Day",
-		"threeD.y.axis.label": "Days of Calendar Year",
+		"threeD.y.axis.label": "Date",
```
The text change does not appear in other languages.

huss commented on src/server/models/Reading.js:
```diff
@@ -475,6 +475,55 @@ class Reading {
+	 * @returns {Promise<{minDate: Moment | null, maxDate: Moment | null}>}
+	 */
+	static async getMeterDataRange(meterID, conn) {
+		const result = await conn.oneOrNone(`
```
OED normally prefers to have an SQL file for queries. Unless there is a reason not to do this then it should be changed. Also, I think using $1 is less desirable and not done in our SQL files.

This also applies to group below.

huss commented on src/server/models/Reading.js:
```diff
@@ -475,6 +475,55 @@ class Reading {
+			FROM readings 
+			WHERE meter_id = $1
+		`, [meterID]);
+		if (result && result.min_date && result.max_date) {
```
My testing found (as I would expect) that if there are not readings so no min/max then you get:

result:  { min_date: null, max_date: null }

Thus, I don't think you need the extra logic for this.

huss commented on src/server/routes/unitReadings.js:
```diff
@@ -466,22 +477,24 @@ function createRouter() {
+				const durationInDays = timeInterval.endTimestamp.diff(timeInterval.startTimestamp, 'days') + 1;
+				// Security: Limit 3D to MAX_3D_DAYS (1095 days = 3 years) to prevent abuse from unauthenticated users.
+				// Frontend auto-adjustment provides UX, but backend must enforce security limits.
+				if (durationInDays > 1095) {
```
As was done on the client, it is nice to make limits a const though it will be a little different due to JS not TS. I think that should be done here, esp. since the value is used more than once.

huss commented on src/client/app/components/ThreeDIntervalControlsComponent.tsx:
```diff
@@ -0,0 +1,177 @@
+			<div style={divTopBottomPadding}>
+				<p style={labelStyle}>
+					{translate('threeD.numDays')}:
+					<TooltipMarkerComponent page='home' helpTextId='help.home.threeD.numDays.tip' />
```
There should be a placeholder in src/client/app/components/TooltipHelpComponent.tsx for this popup help.

huss reviewed (changes requested):
Thanks to @hmhngx for the updated code. There are comments from the past that I think are unresolved & I made some new comments based on the changes. As indicated, it might be good to discuss the big picture ones to decide how they will be resolved. Also, there are three failing tests due to changed in 3D that need to be fixed so they pass.

hmhngx committed:
Move data range SQL queries to separate files

- Create get_meter_data_range.sql and get_group_data_range.sql
- Use named parameters (${meterID}, ${groupID}) instead of $1
- Simplify null check logic using optional chaining

hmhngx committed:
Add MAX_3D_DAYS constant to unitReadings routes

- Define MAX_3D_DAYS = 1095 at module level
- Replace hardcoded 1095 values with constant
- Improves maintainability and consistency

hmhngx committed:
Remove 30-day default from DateRangeComponent

- Allow date range to remain unbounded when queryTimeInterval is unbounded
- 3D component handles its own date range calculation using DEFAULT_3D_DAYS
- Fixes mismatch between displayed date range and actual 3D data range

hmhngx committed:
Remove unused queryTimeInterval from selectThreeDQueryArgs

- queryTimeInterval was in dependencies but not used in selector
- Selector uses threeDInterval instead for 3D date range

hmhngx committed:
Fix translations and add 3D numDays help text

- Fix alphabetical order of 2.years, 3.years, 6.months in French and Spanish
- Update y-axis label from 'Days of Calendar Year' to 'Date' in French and Spanish
- Add help text placeholder for help.home.threeD.numDays.tip

hmhngx committed:
Fix unsaved changes detection for CSV upload fields

- Compare useMeterZone and warnOnCumulativeReset against defaults
- Previously compared values against themselves, always returning false
- Now correctly detects when these fields differ from defaults

hmhngx committed:
Fix enum value mismatch in disable_checks migration

- Change 'reject_disable' to 'reject_disabled' in migration files
- Matches TypeScript/JavaScript code and main enum definition
- Prevents constraint violations when migrating existing data
